File: lib/controllers/auth_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../services/api_service.dart';
import '../services/auth_service.dart';

enum UserRole { patient, admin }

class AuthState {
  const AuthState({
    this.identifier = '',
    this.password = '',
    this.identifierError,
    this.passwordError,
    this.isLoading = false,
    this.isPasswordVisible = false,
    this.selectedRole = UserRole.patient,
    this.errorMessage,
    this.user,
  });

  final String identifier;
  final String password;
  final String? identifierError;
  final String? passwordError;
  final bool isLoading;
  final bool isPasswordVisible;
  final UserRole selectedRole;
  final String? errorMessage;
  final Map<String, dynamic>? user;
}

class AuthController extends StateNotifier<AuthState> {
  AuthController([this._authService]) : super(const AuthState());

  final AuthService? _authService;

  void selectRole(UserRole role) {
    state = AuthState(
      identifier: state.identifier,
      password: state.password,
      identifierError: state.identifierError,
      passwordError: state.passwordError,
      isLoading: state.isLoading,
      isPasswordVisible: state.isPasswordVisible,
      selectedRole: role,
      errorMessage: state.errorMessage,
      user: state.user,
    );
  }

  void setIdentifier(String value) {
    state = AuthState(
      identifier: value,
      password: state.password,
      passwordError: state.passwordError,
      isLoading: state.isLoading,
      isPasswordVisible: state.isPasswordVisible,
      selectedRole: state.selectedRole,
      errorMessage: null,
      user: state.user,
    );
  }

  void setPassword(String value) {
    state = AuthState(
      identifier: state.identifier,
      password: value,
      identifierError: state.identifierError,
      isLoading: state.isLoading,
      isPasswordVisible: state.isPasswordVisible,
      selectedRole: state.selectedRole,
      errorMessage: state.errorMessage,
      user: state.user,
    );
  }

  void togglePasswordVisibility() {
    state = AuthState(
      identifier: state.identifier,
      password: state.password,
      identifierError: state.identifierError,
      passwordError: state.passwordError,
      isLoading: state.isLoading,
      isPasswordVisible: !state.isPasswordVisible,
      selectedRole: state.selectedRole,
      errorMessage: state.errorMessage,
      user: state.user,
    );
  }

  bool validate() {
    final identifierError = state.identifier.trim().isEmpty
        ? 'Medical ID or Email is required'
        : null;
    final passwordError = state.password.isEmpty
        ? 'Password is required'
        : state.password.length < 6
        ? 'Password must be at least 6 characters'
        : null;
    state = AuthState(
      identifier: state.identifier,
      password: state.password,
      identifierError: identifierError,
      passwordError: passwordError,
      isLoading: state.isLoading,
      isPasswordVisible: state.isPasswordVisible,
      selectedRole: state.selectedRole,
      errorMessage: state.errorMessage,
      user: state.user,
    );
    return identifierError == null && passwordError == null;
  }

  Future<bool> signIn() async {
    if (!validate()) return false;
    state = AuthState(
      identifier: state.identifier,
      password: state.password,
      isLoading: true,
      isPasswordVisible: state.isPasswordVisible,
      selectedRole: state.selectedRole,
      errorMessage: null,
      user: state.user,
    );
    if (_authService == null) {
      await Future<void>.delayed(const Duration(milliseconds: 700));
      state = AuthState(
        identifier: state.identifier,
        password: state.password,
        isPasswordVisible: state.isPasswordVisible,
        selectedRole: state.selectedRole,
      );
      return true;
    }
    try {
      final session = await _authService.login(
        state.identifier,
        state.password,
      );
      state = AuthState(
        identifier: state.identifier,
        password: state.password,
        isPasswordVisible: state.isPasswordVisible,
        selectedRole: session.role == 'admin' || session.role == 'nurse'
            ? UserRole.admin
            : UserRole.patient,
        user: session.user,
      );
      return true;
    } on ApiException catch (error) {
      state = AuthState(
        identifier: state.identifier,
        password: state.password,
        isPasswordVisible: state.isPasswordVisible,
        selectedRole: state.selectedRole,
        errorMessage: error.message,
      );
      return false;
    }
  }

  Future<void> logout() async {
    await _authService?.logout();
    state = const AuthState();
  }
}

final authControllerProvider = StateNotifierProvider<AuthController, AuthState>(
  (ref) => AuthController(ref.watch(authServiceProvider)),
);

File: src/controllers/auth.controller.js
const jwt = require('jsonwebtoken');
const mongoose = require('mongoose');

const env = require('../config/env');
const User = require('../models/User');
const PatientProfile = require('../models/PatientProfile');

const publicUser = user => ({
  id: user._id,
  medicalId: user.medicalId,
  email: user.email,
  role: user.role,
  patientProfileId: user.patientProfileId,
  displayName: user.displayName,
  isActive: user.isActive
});

const validationError = (message, details = {}) => {
  const error = new Error(message);
  error.statusCode = 400;
  error.code = 'VALIDATION_ERROR';
  error.details = details;
  return error;
};

const signAccessToken = user => jwt.sign(
  { sub: user._id.toString(), role: user.role },
  env.jwtAccessSecret,
  { expiresIn: '15m', issuer: 'renalflow-api' }
);

const register = async (req, res, next) => {
  const session = await mongoose.startSession();

  try {
    const {
      medicalId,
      email,
      password,
      role = 'patient',
      displayName,
      patientProfile
    } = req.body;

    if (!password || password.length < 8) {
      throw validationError('Password must be at least 8 characters long.');
    }
    if (!displayName || (!medicalId && !email)) {
      throw validationError('displayName and medicalId or email are required.');
    }
    if (!['patient', 'nurse', 'admin'].includes(role)) {
      throw validationError('role must be patient, nurse, or admin.');
    }
    if (role === 'patient' && !patientProfile) {
      throw validationError('patientProfile is required for patient registration.');
    }

    let createdUser;
    await session.withTransaction(async () => {
      const users = await User.create([{
        medicalId,
        email,
        passwordHash: password,
        role,
        displayName
      }], { session });
      createdUser = users[0];

      if (role === 'patient') {
        const profiles = await PatientProfile.create([{
          ...patientProfile,
          patientId: medicalId,
          userId: createdUser._id,
          fullName: patientProfile.fullName || displayName,
          phone: patientProfile.phone || patientProfile.emergencyContact?.phone || 'Not provided',
          bloodGroup: patientProfile.bloodGroup || 'Unknown'
        }], { session });
        createdUser.patientProfileId = profiles[0]._id;
        await createdUser.save({ session });
      }
    });

    res.status(201).json({
      success: true,
      data: { user: publicUser(createdUser) },
      requestId: req.requestId
    });
  } catch (error) {
    if (error.code === 11000) {
      error.statusCode = 409;
      error.code = 'DUPLICATE_USER';
      error.message = 'A user with that medical ID or email already exists.';
    }
    next(error);
  } finally {
    await session.endSession();
  }
};

const login = async (req, res, next) => {
  try {
    const { identifier, medicalId, email, password } = req.body;
    const loginIdentifier = (identifier || medicalId || email || '').trim();

    if (!loginIdentifier || !password) {
      throw validationError('identifier and password are required.');
    }

    const query = loginIdentifier.includes('@')
      ? { email: loginIdentifier.toLowerCase() }
      : { medicalId: loginIdentifier.toUpperCase() };
    const user = await User.findOne(query).select('+passwordHash');

    if (!user || !user.isActive || !(await user.comparePassword(password))) {
      const error = new Error('Invalid credentials.');
      error.statusCode = 401;
      error.code = 'INVALID_CREDENTIALS';
      throw error;
    }

    res.json({
      success: true,
      data: {
        user: publicUser(user),
        accessToken: signAccessToken(user),
        tokenType: 'Bearer',
        expiresIn: '15m'
      },
      requestId: req.requestId
    });
  } catch (error) {
    next(error);
  }
};

const me = (req, res) => {
  res.json({
    success: true,
    data: { user: publicUser(req.user) },
    requestId: req.requestId
  });
};

module.exports = { register, login, me };